From e563a5de1beaaf51bc84b36e702522e7c27a0f22 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:05:32 -0400
Subject: [usb] Revert: xhci: Avoid infinite loop when sg urb requires too many trbs

Message-id: <1410887463-20674-261-git-send-email-dzickus@redhat.com>
Patchwork-id: 95520
O-Subject: [RHEL7 PATCH 260/591] Revert "xhci: Avoid infinite loop when sg urb requires too many trbs"
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit 9cf00d91708221ff2d8a11143315f7ebab8d5da8
Author: Sarah Sharp <sarah.a.sharp@linux.intel.com>
Date:   Fri Jan 31 11:51:59 2014 -0800

    Revert "xhci: Avoid infinite loop when sg urb requires too many trbs"

    This reverts commit d6c9ea9069af684358efedcaf2f2f687f51c58ee.

    We are ripping out commit 35773dac5f862cb1c82ea151eba3e2f6de51ec3e "usb:
    xhci: Link TRB must not occur within a USB payload burst" because it's a
    hack that caused regressions in the usb-storage and userspace USB
    drivers that use usbfs and libusb.  This commit attempted to fix the
    issues with that patch.

    Signed-off-by: Sarah Sharp <sarah.a.sharp@linux.intel.com>
    Cc: stable@vger.kernel.org # 3.12
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/host/xhci-ring.c b/drivers/usb/host/xhci-ring.c
index a0b248c..09b2b55 100644
--- a/drivers/usb/host/xhci-ring.c
+++ b/drivers/usb/host/xhci-ring.c
@@ -3000,7 +3000,7 @@ static int prepare_ring(struct xhci_hcd *xhci, struct xhci_ring *ep_ring,
    if (num_trbs >= TRBS_PER_SEGMENT) {
     xhci_err(xhci, "Too many fragments %d, max %d\n",
       num_trbs, TRBS_PER_SEGMENT - 1);
-    return -EINVAL;
+    return -ENOMEM;
    }
 
    nop_cmd = cpu_to_le32(TRB_TYPE(TRB_TR_NOOP) |
-- 
1.7.1